src/workerd/util/sqlite.h

```diff
@@ -522,11 +544,58 @@ class SqliteDatabase::Query final: private ResetListener {
     }
   }
 
+  class QueryEvent {
+   public:
+    explicit QueryEvent(Query& q)
+        : observer(q.db.sqliteObserver),
+          dbWalSizeBefore(observer.getDbWalSize()),
+          startTime(observer.now()) {}
+
+    ~QueryEvent() noexcept(false) {
+      uint64_t dbWalSizeAfter = observer.getDbWalSize();
+      uint64_t dbWalBytesWritten = (dbWalSizeAfter - dbWalSizeBefore);
+      kj::Duration queryLatency = observer.now() - startTime;
+
+      observer.reportQueryEvent(kj::mv(queryStatement), rowsRead, rowsWritten, queryLatency,
+          dbWalBytesWritten, queryResult, isInternalQuery, kj::mv(queryErrorDescription));
+    }
+
+    void setQueryEventStats(uint64_t rowsRead, uint64_t rowsWritten, bool isInternalQuery) {
+      this->rowsRead = rowsRead;
+      this->rowsWritten = rowsWritten;
+      this->isInternalQuery = isInternalQuery;
+    }
+
+    void setQueryStatement(kj::String queryStatement) {
+      this->queryStatement = kj::mv(queryStatement);
+    }
+
+    void setQueryErrorDescription(kj::String queryErrorDescription) {
+      this->queryErrorDescription = kj::mv(queryErrorDescription);
+    }
+
+    void setQueryResult(int res) {
+      queryResult = res;
+    }
+
+   private:
+    SqliteObserver& observer;
+    kj::Maybe<kj::String> queryStatement = kj::none;
+    bool isInternalQuery;
```

Comment:
I think in some cases the compiler is allowed to leave primitive fields like `int` and `bool` uninitialized, if there is not an explicit value given in either the constructor or a member initializer. So it might not be a bad idea to initialize `isInternalQuery = false` here.